File: SampleNHAppObjC/SampleNHAppObjC/SetupViewController.m
//----------------------------------------------------------------
//  Copyright (c) Microsoft Corporation. All rights reserved.
//----------------------------------------------------------------

#import "SetupViewController.h"
#import "NotificationsTableViewController.h"

static NSString *const kNHMessageReceived = @"MessageReceived";
static void * const SetupViewControllerKVOContext = (void*)&SetupViewControllerKVOContext;

@interface SetupViewController ()

@end

@implementation SetupViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [[ANHNotificationHub sharedInstance] addObserver:self forKeyPath:@"pushChannel" options:NSKeyValueObservingOptionNew context:SetupViewControllerKVOContext];
    
    // Do any additional setup after loading the view.
    self.tags = [ANHNotificationHub sharedInstance].tags;
    
    self.addNewTagTextField.delegate = self;
    self.tagsTable.delegate = self;
    self.tagsTable.dataSource = self;
    [self.tagsTable reloadData];
    self.userId.delegate = self;
    
    self.deviceTokenLabel.text = [ANHNotificationHub sharedInstance].pushChannel;
    self.installationIdLabel.text = [ANHNotificationHub sharedInstance].installationId;
    self.userId.text = [ANHNotificationHub sharedInstance].userId;
    
    self.notificationsTableView = (NotificationsTableViewController*) [[(UINavigationController*)[[self.tabBarController viewControllers] objectAtIndex:1] viewControllers] objectAtIndex:0];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didReceivePushNotification:) name:kNHMessageReceived object:nil];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kNHMessageReceived object:nil];
    
    [[ANHNotificationHub sharedInstance] removeObserver:self forKeyPath:@"pushChannel" context:SetupViewControllerKVOContext];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                  ofObject:(id)object
                    change:(NSDictionary *)change
                   context:(void *)context {
    if (context != SetupViewControllerKVOContext) {
        [super observeValueForKeyPath:keyPath
                             ofObject:object
                               change:change
                              context:context]; return;
        
    }
    
    if ([keyPath isEqual: @"pushChannel"]) {
        self.deviceTokenLabel.text = [ANHNotificationHub sharedInstance].pushChannel;
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    // Hide the keyboard.
    [textField resignFirstResponder];
    return YES;
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    if(textField.tag == 0) {
        [ANHNotificationHub sharedInstance].userId = textField.text;
    } else if (![textField.text isEqual: @""]) {
        [[ANHNotificationHub sharedInstance] addTag:textField.text];
        self.tags = [ANHNotificationHub sharedInstance].tags;
        textField.text = @"";
        [self.tagsTable reloadData];
    }
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    NSString *cellIdentifier = @"TagCell";
    
    TagTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier forIndexPath:indexPath];
    
    cell.tagLabel.text = self.tags[indexPath.row];
    
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.tags count];
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [[ANHNotificationHub sharedInstance] removeTag:self.tags[indexPath.row]];
        self.tags = [ANHNotificationHub sharedInstance].tags;
        [tableView deselectRowAtIndexPath:indexPath animated:YES];
        [self.tagsTable reloadData];
    }
}

- (void)didReceivePushNotification:(NSNotification *)notification {
    ANHNotificationHubMessage *message = [notification.userInfo objectForKey:@"message"];
    
    [self.notificationsTableView addNotification:message];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:message.title
                         message:message.body
                  preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alertController animated:YES completion:nil];
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [alertController dismissViewControllerAnimated:YES completion: nil];
    });
}

@end

File: SampleNHAppSwift/SampleNHAppSwift/SetupViewController.swift
//----------------------------------------------------------------
//  Copyright (c) Microsoft Corporation. All rights reserved.
//----------------------------------------------------------------

import UIKit

class SetupViewController: UIViewController, UITextFieldDelegate, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var deviceTokenLabel: UILabel!
    @IBOutlet weak var installationIdLabel: UILabel!
    @IBOutlet weak var addNewTagTextField: UITextField!
    @IBOutlet weak var tagsTable: UITableView!
    @IBOutlet weak var userId: UITextField!
    
    var tags = MSNotificationHub.getTags()
    var notificationsTableView:NotificationsTableViewController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        addNewTagTextField.delegate = self
        tagsTable.delegate = self
        tagsTable.dataSource = self
        tagsTable.reloadData()
        userId.delegate = self
        
        deviceTokenLabel.text = MSNotificationHub.getPushChannel()
        installationIdLabel.text = MSNotificationHub.getInstallationId()
        userId.text = MSNotificationHub.getUserId()
        
        notificationsTableView = (self.tabBarController?.viewControllers?[1] as! UINavigationController).viewControllers[0] as? NotificationsTableViewController
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.didReceivePushNotification), name: Notification.Name("MessageReceived"), object: nil);
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: Notification.Name("MessageReceived"), object: nil)
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        // Hide the keyboard.
        textField.resignFirstResponder()
        return true
    }
    
    func textFieldDidEndEditing(_ textField: UITextField) {
        if(textField.tag == 0) {
            MSNotificationHub.setUserId(textField.text!)
        } else if (textField.text != "") {
            MSNotificationHub.addTag(textField.text!)
            tags = MSNotificationHub.getTags()
            textField.text = ""
            tagsTable.reloadData()
        }
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tags.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "TagCell"
        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? TagTableViewCell  else {
            fatalError("The dequeued cell is not an instance of TagCell.")
        }
        
        cell.tagLabel.text = tags[indexPath.row]
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
      if editingStyle == .delete {

        MSNotificationHub.removeTag(tags[indexPath.row])
        tags = MSNotificationHub.getTags()
        tableView.deleteRows(at: [indexPath], with: .automatic)
        tagsTable.reloadData()
      }
    }
    
    @objc func didReceivePushNotification(notification: Notification) {
        let message = notification.userInfo!["message"] as! MSNotificationHubMessage;
        NSLog("Received notification: %@; %@", message.title ?? "<nil>", message.body)
        notificationsTableView?.addNotification(message);
        
        let alertController = UIAlertController(title: message.title, message: message.body, preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "OK", style: .cancel))
        self.present(alertController, animated: true, completion: nil)
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            alertController.dismiss(animated: true, completion: nil)
        }
    }
    
}

